### DISCUSSION TITLE
How to clean up unmerged commits

### DISCUSSION THREAD

**Discussion by kt-brandi-larue** [community/community#110548] [New to GitHub]
Hi! 

I am new here and was wondering how to clean up, delete or archive unmerged commits in a branch. I do not want to delete the branch but only want to get rid of the unmerged commits and start fresh on an existing branch.

**Comment by haroonmaqbool** [comment]
To clean up unmerged commits on a branch without deleting it:

Use these commands:

**1. Checkout the Branch:**  
```
git checkout <branch_name>
```

**2. Reset the Branch:**  
```
git reset --hard <commit_hash>
```

**3. Force Push:**  
```
git push origin <branch_name> --force
```

Replace `<branch_name>` with your branch's name and `<commit_hash>` with the commit before the unmerged commits.

**Reply by koeppel** [reply]
Can also reset the branch by typing

```sh
git reset --hard origin/<main_branch_name>
```

You might also want to try --soft instead of --hard so that you can filter out which changes you want to keep and which you want to drop.

**Comment by Lanrad23** [comment]
To completely remove the unmerged commits and start fresh (as if the branch was just created), you can reset it to the commit where it diverged from the main branch or to any specific commit you want to go back to. Use the git log command to find the commit ID you want to reset to.

**Comment by GustaFranz** [comment]
Think of a Git branch as a movable bookmark in your project's history. The unwanted commits are pages after the point where you want the bookmark to be. Resetting the branch moves that bookmark backward; it does not require deleting the branch itself.

Before changing anything, create a backup pointer:

```bash
git switch your-branch
git branch backup/your-branch-before-reset
```

The second command creates another branch pointing to the current commit. If you reset to the wrong place, your original work is still reachable through the backup branch.

Next, inspect the history:

```bash
git log --oneline --graph --decorate
```

Find the last commit you want to keep, then move the current branch to it:

```bash
git reset --hard <commit-to-keep>
```

`reset` moves the branch pointer. `--hard` also makes the working directory and staging area match that commit, so uncommitted changes will be discarded. If you need to preserve local changes, stop and use `--soft` or `--mixed` instead.

If the unwanted commits were already pushed, the local and remote branches will now disagree. Update the remote with:

```bash
git push --force-with-lease origin your-branch
```

`--force-with-lease` is safer than `--force`: it refuses to overwrite the remote branch if someone else updated it since your last fetch.

If this is a shared branch, coordinate with the other contributors first because rewriting its history changes the commit IDs they have locally.

After confirming that everything is correct, the backup branch can be removed:

```bash
git branch -d backup/your-branch-before-reset
```
